java/maven/src/org/netbeans/modules/maven/queries/MavenPrimingReloadImplementation.java

```diff
@@ -0,0 +1,251 @@
+            context.saveLoadContext(this);
+            return CompletableFuture.completedFuture(null);
+        } 
+        
+        ActionProvider ap = project.getLookup().lookup(ActionProvider.class);
+        if (!ap.isActionEnabled(ActionProvider.COMMAND_PRIME, Lookup.EMPTY)) {
+            return CompletableFuture.completedFuture(null);
+        }
+
+        Collection<Artifact> faked = MavenProjectCache.getFakedArtifacts(p);
+        Set<String> gavs = new HashSet<>();
+        faked.forEach(a -> gavs.add(artifactGav(a)));
+        String parentGav = artifactGav(p.getParentArtifact());
+        
+        CompletableFuture<ProjectStateData> future = new CompletableFuture<>();
+        if (gavs.contains(parentGav)) {
+            if (request.getMinQuality().isWorseThan(ProjectReload.Quality.BROKEN)) {
+                future.complete(createStateData(p));
+            } else if (request.isOfflineOperation()) {
+                PartialLoadException ex = new PartialLoadException(createStateData(p),
+                     Bundle.ERR_ParentPomMissing(ProjectUtils.getInformation(project).getDisplayName(), parentGav), 
+                    new ProjectOperationException(project, ProjectOperationException.State.OFFLINE, 
+                    Bundle.ERR_UnprimedInOfflineMode(ProjectUtils.getInformation(project).getDisplayName()))
+                );
+                future.completeExceptionally(ex);
+                return future;
+            }
+        }
+        if (request.getMinQuality().isWorseThan(ProjectReload.Quality.RESOLVED)) {
+            future.complete(createStateData(p));
+            return future;
+        } else if (request.isOfflineOperation()) {
+            PartialLoadException ex = new PartialLoadException(createStateData(p),
+                 Bundle.ERR_ParentPomMissing(ProjectUtils.getInformation(project).getDisplayName(), parentGav), 
+                new ProjectOperationException(project, ProjectOperationException.State.OFFLINE, 
+                    Bundle.ERR_UnprimedInOfflineMode(ProjectUtils.getInformation(project).getDisplayName()))
+            );
+            future.completeExceptionally(ex);
+            return future;
+        }
+        
+        synchronized (this) {
+            if (!lc.firstRun && fakeArtifactNames.containsAll(gavs) && !request.isForceReload()) {
+                // no point in running priming build again, when the artifacts are known to be broken.
+                future.complete(createStateData(p));
+                return future;
+            }
+            fakeArtifactNames = gavs;
+        }
+
+        lc.firstRun = false;
+
+        class C extends AtomicReference<Cancellable> implements Consumer<Cancellable> {
+            @Override
+            public void accept(Cancellable t) {
+                set(t);
+            }
+        }
+        C c = new C();
+        context.setCancellable(() -> {
+            Cancellable h = c.get();
+            return h != null && h.cancel();
+        });
```

Comment:
so what is `C c`,  a `CancelSignalDelagator`?

something like this perhaps:
```java
        class CancelSignalDelegator implements Consumer<Cancellable>, Cancellable {
            private final AtomicReference<Cancellable> ref = new AtomicReference<>();
            @Override
            public void accept(Cancellable other) {
                ref.set(other);
            }
            @Override
            public boolean cancel() {
                Cancellable other = ref.get();
                return other != null && other.cancel();
            }
        }
        Cancellable csd = new CancelSignalDelegator();
        context.setCancellable(csd);
```
`AtomicReference` moved to a impl detail, rest is now self documenting by giving things names ;)

Reply:
Accepted in 4c091822130a